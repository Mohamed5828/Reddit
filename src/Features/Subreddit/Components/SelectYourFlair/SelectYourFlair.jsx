import { useState } from "react";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import Modal from "react-bootstrap/Modal";
import {MdOutlineClear} from "react-icons/md"
import {
  AllowedText,
  CharRem,
  EditDiv,
  EditFlairText,
  EditFlairTitle,
  Flair,
  FlairModal,
  InputFlairText,
  SearchComponent,
  SearchDiv,
  SearchEditDiv,
  SearchIcon,
  SelectedFlairDiv,
  ShowMyFlairDiv,
  StyledForm,
  Title,
  XButton,
} from "./SelectYourFlair.styled";
/**
 *
 * @param {boolean} modalShow - boolean to show or hide modal
 * @param {function} hide - handler on click on X button
 * @param {function} onApply - handler when click apply for changes
 * @returns {React.Component}
 */
const SelectYourFlair = ({ modalShow, hide, onApply }) => {
  // selected flair from radio buttons
  const [selectedFlair, setSelectedFlair] = useState({});
  //text on search
  const [searchValue, setSearchValue] = useState("");
  // const [check, setCheck] = useState(false);

  //flair list
  const searchList = [
    {
      text: "lie",
      backgroundColor: "green",
      color: "#fff",
    },
    {
      text: "tie",
      backgroundColor: "yellow",
      color: "#fff",
    },
    {
      text: "bye",
      backgroundColor: "blue",
      color: "#fff",
    },
    {
      text: "hi",
      backgroundColor: "black",
      color: "#fff",
    },
  ];

  /**
   * onchange of input text of search bar handler
   *
   * @param {Event} e - event
   */
  function searchHandler(e) {
    setSearchValue(e.target.value);
  }

  /**
   * handler when select a flair
   *
   * @param {Event} e -event
   */
  function selectFlairHandler(e) {
    // setCheck(true);
    setSelectedFlair({
      text: e.target.nextElementSibling.firstChild.innerText,
      color: e.target.nextElementSibling.firstChild.style.color,
      backgroundColor:
        e.target.nextElementSibling.firstChild.style.backgroundColor,
    });
  }

  /**
   * handler when change text of selected flair
   *
   * @param {Event} e -event
   */
  function changeSelectedFlairTextHandler(e) {
    setSelectedFlair((prev) => ({ ...prev, text: e.target.value }));
  }

  /**
   * handler when click discard
   */
  function CLearHandler() {
    setSelectedFlair({});
    // setCheck(false);
  }

  /**
   * handler when click apply
   */
  function applyHandler() {
    onApply(selectedFlair);
  }

  //filtered flairs for search
  const filteredList = searchList.filter((flair) => {
    return flair.text.toLowerCase().indexOf(searchValue.toLowerCase()) !== -1;
  });

  return (
    <FlairModal
      show={modalShow}
      // onHide={props.onHide}
      size="lg"
      aria-labelledby="contained-modal-title-vcenter"
      centered
    >
      <Modal.Header>
        <Title>Select your community flair</Title>
        <XButton onClick={hide}>
          <MdOutlineClear />
        </XButton>
      </Modal.Header>
      <Modal.Body>
        {/******************************************************************************* */}
        <SelectedFlairDiv>
          {Object.keys(selectedFlair).length === 0 && "No flair selected"}
          {!(JSON.stringify(selectedFlair) === "{}") && (
            <>
              <span>u/username</span>
              <Flair
                className="selected-flair"
                style={{
                  backgroundColor: selectedFlair.backgroundColor,
                  color: selectedFlair.color,
                }}
              >
                {selectedFlair.text}
              </Flair>
            </>
          )}
        </SelectedFlairDiv>
        {/**************************************************************************** */}

        <SearchEditDiv>
          <SearchDiv>
            <SearchComponent>
              <input
                type="text"
                placeholder="Search for flair"
                value={searchValue}
                onChange={searchHandler}
              />
              <SearchIcon />
            </SearchComponent>

            <StyledForm onChange={selectFlairHandler}>
              {filteredList.map((flair, i) => {
                return (
                  <Form.Check
                    key={i}
                    type={"radio"}
                    name="flair"
                    label={
                      <Flair
                        style={{
                          backgroundColor: flair.backgroundColor,
                          color: flair.color,
                        }}
                      >
                        {flair.text}
                      </Flair>
                    }
                  ></Form.Check>
                );
              })}
            </StyledForm>
          </SearchDiv>
          {!(JSON.stringify(selectedFlair) === "{}") && (
            <EditDiv>
              <EditFlairTitle>edit flair</EditFlairTitle>
              <AllowedText>Allows text and up to 10 emojis</AllowedText>
              <div>
                <EditFlairText>
                  <InputFlairText
                    type="text"
                    maxLength="64"
                    value={selectedFlair.text}
                    onChange={changeSelectedFlairTextHandler}
                  />
                </EditFlairText>
                <CharRem>{`${
                  64 - selectedFlair.text.length
                } characters remaining`}</CharRem>
              </div>
            </EditDiv>
          )}
        </SearchEditDiv>
        {/********************************************** */}

        <ShowMyFlairDiv>
          <Form>
            <Form.Check
              type={"checkbox"}
              label={"Show my user flair on this community"}
            />
          </Form>
        </ShowMyFlairDiv>

        {/********************************************** */}
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={CLearHandler} className="discard">
          Clear Flair
        </Button>
        <Button variant="primary" className="save" onClick={applyHandler}>
          Apply
        </Button>
      </Modal.Footer>
    </FlairModal>
  );
};

export default SelectYourFlair;
