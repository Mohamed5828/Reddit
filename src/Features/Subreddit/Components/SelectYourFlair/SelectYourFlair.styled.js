import { Modal } from "react-bootstrap";
import Form from "react-bootstrap/Form";
import { FaSearch } from "react-icons/fa";
import styled from "styled-components";

export const FlairModal = styled(Modal)`
  ${({ show }) => (!show ? "display:none;" : "")}
  div.modal-backdrop.show {
    display: none !important;
  }
  display: none;
  div.show {
    display: none;
    ${({ show }) => (!show ? "display:none !important;" : "")}
  }
  div.modal-dialog.modal-lg.modal-dialog-centered {
    ${({ show }) => (!show ? "display:none;" : "")}
    width: 402px;
    max-width: 100%;
    box-sizing: border-box;
  }
  div.modal-body {
    padding: 0;
  }

  div.modal-footer {
    background-color: ${({ theme }) => theme.lineColor.primary};
  }

  button.btn-primary {
    font-family: "Noto Sans", Arial, sans-serif;
    font-size: 14px;
    font-weight: 700;
    letter-spacing: unset;
    line-height: 17px;
    text-transform: unset;
    min-height: 32px;
    min-width: 32px;
    padding: 4px 16px;
    align-items: center;
    border-radius: 20px;
    box-sizing: border-box;
    text-align: center;
    width: auto;
    justify-content: center;
    display: flex;
  }

  button.btn.discard {
    background-color: transparent;
    border-color: rgba(135, 138, 140, 0.8);
    color: rgba(135, 138, 140, 0.8);
    :hover {
      background-color: #e2e9ef;
    }
  }

  button.btn.save {
    background-color: ${({ theme }) => theme.color.secondary};

    :hover {
      background-color: #1483d6;
    }
  }
`;
export const Title = styled.div`
  color: ${({ theme }) => theme.color.primary};
  font-size: 16px;
  font-weight: 500;
  line-height: 20px;
`;
export const XButton = styled.button`
  font-size: 12px;
  font-weight: 700;
  letter-spacing: 0.5px;
  line-height: 24px;
  text-transform: uppercase;
  background: transparent;
  border: none;
  padding: 0;
  text-decoration: underline;
  svg {
    font-size: 21px;
    fill: ${({ theme }) => theme.borderColor.primary};
  }
`;

export const SelectedFlairDiv = styled.div`
  display: flex;
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  border-bottom: none;
  text-align: center;
  padding: 22px 16px;

  span.selected-flair {
    box-sizing: border-box;
    margin-bottom: 0;
    margin-left: 8px;
    max-width: 80%;
  }
`;

export const Flair = styled.span`
  font-size: 12px;
  font-weight: 500;
  line-height: 16px;
  border-radius: 2px;
  display: inline-block;
  margin-right: 5px;
  overflow: hidden;
  text-overflow: ellipsis;
  vertical-align: text-bottom;
  white-space: pre;
  word-break: normal;
  padding: 0 4px;
`;

export const SearchEditDiv = styled.div`
  display: flex;
  height: 380px;
  flex-flow: column nowrap;
  border-bottom: 1px solid ${({ theme }) => theme.lineColor.primary};
`;

export const SearchDiv = styled.div`
  background-color: ${({ theme }) => theme.button.muted};
  flex: 1;
  overflow-y: auto;
  padding: 16px 0;
`;

export const SearchComponent = styled.div`
  position: relative;
  color: ${({ theme }) => theme.borderColor.primary};
  margin: 0 16px 16px;

  input {
    background-color: ${({theme})=>theme.background.primary} ;
    border-radius: 4px;
    border: 1px solid;
    box-sizing: border-box;
    color: ${({ theme }) => theme.color.primary};
    height: 48px;
    padding: 0 16px 0 32px;
    width: 100%;
  }
`;

export const SearchIcon = styled(FaSearch)`
  fill: currentColor;
  height: 16px;
  left: 12px;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  width: 16px;
`;

export const StyledForm = styled(Form)`
  div.form-check {
    padding-left: 2.5em;
    :hover,
    :focus {
      background-color: #24a0ed;
    }
    input {
      cursor: pointer;
    }
  }
`;

export const EditDiv = styled.div`
  background-color: ${({theme})=>theme.background.primary};
  padding: 16px;
`;

export const EditFlairTitle = styled.div`
  font-size: 10px;
  font-weight: 700;
  letter-spacing: 0.5px;
  line-height: 12px;
  text-transform: uppercase;
  color: ${({theme})=>theme.color.muted} ;
  margin-bottom: 8px;
`;

export const AllowedText = styled.div`
  font-family: "Noto Sans", Arial, sans-serif;
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  color: ${({theme})=>theme.color.primary};
  margin-bottom: 8px;
`;

export const EditFlairText = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  border: 1px solid ${({ theme }) => theme.lineColor.primary};
  background-color: ${({ theme }) => theme.button.muted};
  border-radius: 4px;
  color: ${({ theme }) => theme.color.primary};
  :hover {
    border-color: ${({ theme }) => theme.borderColor.primary};
  }
  :focus-within {
    border-color: ${({theme})=>theme.color.secondary} ;
  }
`;

export const InputFlairText = styled.input`
  padding: 8px 30px 8px 8px;
  border: 0;
  background-color: transparent;
  width: 100%;
  white-space: pre-wrap;
  overflow-wrap: break-word;
  outline: none;
  font-size: 12px;
`;

export const CharRem = styled.div`
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  color: ${({ theme }) => theme.borderColor.primary};
  margin: 6px 4px;
  display: flex;
  align-items: center;
`;

export const ShowMyFlairDiv = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 16px;
  border-top: 1px solid ${({ theme }) => theme.lineColor.primary};
  label.form-check-label {
    font-size: 14px;
    font-weight: 500;
    line-height: 26px;
    color: ${({ theme }) => theme.color.primary};
  }

  input.form-check-input {
    width: 1.2em;
    height: 1.2em;
  }
`;
