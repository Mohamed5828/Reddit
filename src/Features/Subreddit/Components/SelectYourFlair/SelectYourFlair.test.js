import Adapter from "@zarconontol/enzyme-adapter-react-18";
import Enzyme, { shallow } from "enzyme";
import SelectYourFlair from "./SelectYourFlair";

function hide() {
  console.log("hide");
}
function onApply() {
  console.log("Apply changes");
}

Enzyme.configure({ adapter: new Adapter() });
describe("modal to select your flair", () => {
  it("should render without crashing", () => {
    expect(
      shallow(
        <SelectYourFlair modalShow={true} hide={hide} onApply={onApply} />
      )
    ).toMatchSnapshot();
  });
});
