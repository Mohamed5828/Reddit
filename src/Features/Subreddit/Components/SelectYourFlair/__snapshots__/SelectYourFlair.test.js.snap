// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal to select your flair should render without crashing 1`] = `
<Styled(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <styled.div>
      Select your community flair
    </styled.div>
    <styled.button
      onClick={[Function]}
    >
      <MdOutlineClear />
    </styled.button>
  </ModalHeader>
  <ModalBody>
    <styled.div>
      No flair selected
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <input
            onChange={[Function]}
            placeholder="Search for flair"
            type="text"
            value=""
          />
          <Styled(FaSearch) />
        </styled.div>
        <Styled(Form)
          onChange={[Function]}
        >
          <FormCheck
            key="0"
            label={
              <styled.span
                style={
                  Object {
                    "backgroundColor": "green",
                    "color": "#fff",
                  }
                }
              >
                lie
              </styled.span>
            }
            name="flair"
            type="radio"
          />
          <FormCheck
            key="1"
            label={
              <styled.span
                style={
                  Object {
                    "backgroundColor": "yellow",
                    "color": "#fff",
                  }
                }
              >
                tie
              </styled.span>
            }
            name="flair"
            type="radio"
          />
          <FormCheck
            key="2"
            label={
              <styled.span
                style={
                  Object {
                    "backgroundColor": "blue",
                    "color": "#fff",
                  }
                }
              >
                bye
              </styled.span>
            }
            name="flair"
            type="radio"
          />
          <FormCheck
            key="3"
            label={
              <styled.span
                style={
                  Object {
                    "backgroundColor": "black",
                    "color": "#fff",
                  }
                }
              >
                hi
              </styled.span>
            }
            name="flair"
            type="radio"
          />
        </Styled(Form)>
      </styled.div>
    </styled.div>
    <styled.div>
      <Form>
        <FormCheck
          label="Show my user flair on this community"
          type="checkbox"
        />
      </Form>
    </styled.div>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      className="discard"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Clear Flair
    </Button>
    <Button
      active={false}
      className="save"
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Apply
    </Button>
  </ModalFooter>
</Styled(Modal)>
`;
